game/scripts/vscripts/items/azazel_tower_defense.lua

```diff
@@ -0,0 +1,158 @@
+    self:GetCaster():IsPositionInRange(location, 144 + self:GetCaster():GetHullRadius())
+  then
+    return UF_FAIL_CUSTOM
+  else
+    return UF_SUCCESS
+  end
+end
+function item_azazel_tower_defense_1:GetCustomCastErrorLocation(location)
+  return "#dota_hud_error_no_buildings_here"
+end
+
+function item_azazel_tower_defense_1:OnSpellStart()
+  local caster = self:GetCaster()
+  local location = self:GetCursorPosition()
+  local building = CreateUnitByName("npc_azazel_tower_defense", location, true, caster, caster:GetOwner(), caster:GetTeam())
+  building:SetOwner(caster)
+  GridNav:DestroyTreesAroundPoint(location, building:GetHullRadius(), true)
+  building:SetOrigin(location)
+  building:RemoveModifierByName('modifier_invulnerable')
+  building:AddNewModifier(building, self, "modifier_defense_tower_construction", {duration = -1})
+  building:AddNewModifier(building, self, "modifier_defense_tower_true_sight", {duration = -1})
+  local charges = self:GetCurrentCharges() - 1
+  if charges < 1 then
+    caster:RemoveItem(self)
+  else
+    self:SetCurrentCharges(charges)
+  end
+end
+
+-- upgrades
+item_azazel_tower_defense_2 = item_azazel_tower_defense_1
+item_azazel_tower_defense_3 = item_azazel_tower_defense_1
+item_azazel_tower_defense_4 = item_azazel_tower_defense_1
+
+--------------------------------------------------------------------------
+-- base modifier
+
+modifier_defense_tower_construction = class(ModifierBaseClass)
+
+local SINK_HEIGHT = 300
+local THINK_INTERVAL = 0.1
+
+function modifier_defense_tower_construction:OnCreated()
+  if IsServer() then
+    local target = self:GetParent()
+    local ab = self:GetAbility()
+    local maxhealth = target:GetMaxHealth() + ab:GetSpecialValueFor("bonus_health")
+    local location = target:GetOrigin()
+    local time = ab:GetSpecialValueFor("construction_time")
+    target:Attribute_SetIntValue("construction_time", time)
+    target:Attribute_SetIntValue("bonus_damage", ab:GetSpecialValueFor("bonus_damage"))
+    target:SetOrigin(GetGroundPosition(location, target) - Vector(0, 0, SINK_HEIGHT))
+    Timers:CreateTimer(0.1, function()
+      ResolveNPCPositions(location, target:GetHullRadius())
+      target:SetMaxHealth(maxhealth)
+      target:SetHealth(maxhealth * 0.01)
+      self:StartIntervalThink(THINK_INTERVAL)
+      self:SetStackCount(math.floor(time / THINK_INTERVAL)) -- `construction_time` should be divisible by `THINK_INTERVAL`!
+    end)
+  end
+end
+function modifier_defense_tower_construction:OnIntervalThink()
+  if IsServer() then
+    local target = self:GetParent()
+    local time = target:Attribute_GetIntValue("construction_time", 10)
+    local count = self:GetStackCount()
+    local location = target:GetOrigin()
+    target:SetOrigin(target:GetOrigin() + Vector(0, 0, SINK_HEIGHT / (time / THINK_INTERVAL)))
+    self:SetStackCount(count - 1)
+    if count < 1 then
+      self:StartIntervalThink(-1)
```

Comment:
I think the modifier should just self-destruct (`self:Destroy()`). And the other stuff that persists after construction like damage put on another modifier.

Reply:
I don't think a load of modifiers is any better. But if you insist, so be it.

Reply:
Renaming it then seems the way to go then, IMHO.

Reply:
Well, there's also the matter that this method requires a bunch of places to check whether construction is done.